module8/Homework8/Homework7/Views/APIListView.swift

```diff
@@ -0,0 +1,90 @@
+/// Copyright (c) 2024 Kodeco Inc.
+/// 
+/// Permission is hereby granted, free of charge, to any person obtaining a copy
+/// of this software and associated documentation files (the "Software"), to deal
+/// in the Software without restriction, including without limitation the rights
+/// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+/// copies of the Software, and to permit persons to whom the Software is
+/// furnished to do so, subject to the following conditions:
+/// 
+/// The above copyright notice and this permission notice shall be included in
+/// all copies or substantial portions of the Software.
+/// 
+/// Notwithstanding the foregoing, you may not use, copy, modify, merge, publish,
+/// distribute, sublicense, create a derivative work, and/or sell copies of the
+/// Software in any work that is designed, intended, or marketed for pedagogical or
+/// instructional purposes related to programming, coding, application development,
+/// or information technology.  Permission for such use, copying, modification,
+/// merger, publication, distribution, sublicensing, creation of derivative works,
+/// or sale is expressly withheld.
+/// 
+/// This project and source code may use libraries or frameworks that are
+/// released under various Open-Source licenses. Use of those libraries and
+/// frameworks are governed by their own individual licenses.
+///
+/// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+/// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+/// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+/// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+/// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+/// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+/// THE SOFTWARE.
+
+import SwiftUI
+
+struct APIListView: View {
+  @ObservedObject var apiStore = APIStore()
+  @Binding var showingErrorView: Bool
+  var showingAPILoadingIndicator = false
+
+  var body: some View {
+    NavigationStack {
+      ZStack {
+        List {
+          ForEach(apiStore.apiDataList) { apiData in
+            NavigationLink(value: apiData) {
+              Text(apiData.name ?? "")
+                .padding(.vertical)
+            }
+          }
+        }
+        if showingAPILoadingIndicator {
+          ProgressView()
+        }
+      }
+      .listStyle(.plain)
+      .navigationDestination(for: APIData.self) { apiData in
+        APIDetailsView(appStore: apiStore, apiData: apiData)
+          .navigationTitle(Text("API Details"))
+      }
+      .navigationTitle(Text("APIs"))
+    }
+    .onAppear {
+      if apiStore.apiDataState != .loaded {
+        Task {
+          await apiStore.readJSON()
```

Comment:
Getting a warning here:
"Passing argument of non-sendable type 'Binding<Float>?' outside of main actor-isolated context may introduce data races"

which is making me think I am not configuring this correctly w/re to async tasks.

Reply:
You don't need that binding in `progress`. It's making things too complicated and that's what the error is about. Any variable marked as `Binding` is not sendable. If a variable is `Binding` there is a two way direct link between the items that use it. Since you've got things on different threads the changes will happen in an undefined order so eventually something will try to update the `Binding` variable when it shouldn't or when another thread is/has updated it. Since each of your stores are `ObservableObjects` consider having your `JSONDataLoading` protocol have a `progress` variable that the stores can adopt and then push up. Binding is for two way communication. You don't have that here.

Reply:
OK good catch. I didnt even realize I still had the Binding in there, since I wasnt using it in the UI.

Interestingly, in the course materials: https://www.kodeco.com/28450876-beginning-networking-with-urlsession/lessons/14 they do show the example of passing the Binding. And checking out the associated github example shows the same warning.

I managed to re-work things using a more direct callback on the ByteLoader, and then using the viewmodel to publish that into SwiftUI more idiomatically (and make sure the published updates are on the MainActor). I do worry that I am blocking the ByteLoader async for-loop w/ the await on the callback function however.

Definitely a lot of subtleties trying to coordinate these updates. Let me know if there is a simpler way.